doc/neps/nep-0050-scalar-promotion.rst

```diff
@@ -0,0 +1,729 @@
+  enable weak promotion.)
+
+
+NumPy scalars could be special
+------------------------------
+
+Many users expect that NumPy scalars should be different from NumPy
+arrays, in that ``np.uint8(3) + 3`` should return an ``int64`` (or Python
+integer), when ``uint8_arr + 3`` preserves the ``uint8`` dtype.
+
+This alternative would be very close to the current behaviour for NumPy scalars
+but it would cement a distinction between arrays and scalars (NumPy arrays
+are "stronger" than Python scalars, but NumPy scalars are not).
+
+Such a distinction is very much possible, however, at this time NumPy will
+often (and silently) convert 0-D arrays to scalars.
+It may thus make sense, to only consider this alternative if we also
+change this silent conversion (sometimes refered to as "decay") behaviour.
+
+
+Handling conversion of scalars when unsafe
+------------------------------------------
+
+Cases such as::
+
+  np.arange(10, dtype=np.uint8) + 1000
+
+should raise an error as per this NEP.  This could be relaxed to give a warning
+or even ignore the "unsafe" conversion which (on all relevant hardware) would
+lead to ``np.uint8(1000) == np.uint8(232)`` being used.
+
+
+Allowing weakly typed arrays
+----------------------------
+
+One problem with having weakly typed Python scalars, but not weakly typed
+arrays is that in many cases ``np.asarray()`` is called indiscriminately on
+inputs.  To solve this issue JAX will consider the result of ``np.asarray(1)``
+also to be weakly typed.
+There are, however, two difficulties with this:
+
+1. JAX noticed that it can be confusing that::
+
+     np.broadcast_to(np.asarray(1), (100, 100))
+
+   is a non 0-D array that "inherits" the weak typing. [2]_
+2. Unlike JAX tensors, NumPy arrays are mutable, so assignment may need to
+   cause it to be strongly typed?
+
+A flag will likely be useful as an implementation detail (e.g. in ufuncs),
+however, as of now we do not expect to have this as user API.
+The main reason is that such a flag may be surprising for users if it is
+passed out as a result from a function, rather than used only very localized.
+
+
+.. admonition:: TODO
+
+    Before accepting the NEP it may be good to discuss this issue further.
+    Libraries may need clearer patterns to "propagate" the "weak" type, this
+    could just be an ``np.asarray_or_literal()`` to preserve Python scalars,
+    or a pattern of calling ``np.result_type()`` before ``np.asarray()``.
+
+
+Keep using value-based logic for Python scalars
+-----------------------------------------------
+
+Some of the main issues with the current logic arise, because we apply it
+to NumPy scalars and 0-D arrays, rather than the application to Python scalars.
+We could thus consider to keep inspecting the value for Python scalars.
+
+We reject this idea on the grounds that it will not remove the surprises
+given earlier::
+
+    np.uint8(100) + 1000 == np.uint16(1100)
+    np.uint8(100) + 200 == np.uint8(44)
```

Comment:
I'd argue that the first one is fine and the second should be `np.uint16(300)` (more details in the comments above).